sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala

```diff
@@ -3488,6 +3488,35 @@ class DataSourceV2SQLSuiteV1Filter
     }
   }
 
+  test("SPARK-36680: Supports Dynamic Table Options for Spark SQL") {
+    val t1 = s"${catalogAndNamespace}table"
+    withTable(t1) {
+      sql(s"CREATE TABLE $t1 (id bigint, data string) USING $v2Format")
+      sql(s"INSERT INTO $t1 VALUES (1, 'a'), (2, 'b')")
+
+      var df = sql(s"SELECT * FROM $t1")
+      var collected = df.queryExecution.optimizedPlan.collect {
+        case scan: DataSourceV2ScanRelation =>
+          assert(scan.relation.options.isEmpty)
+      }
+      assert (collected.size == 1)
+      checkAnswer(df, Seq(Row(1, "a"), Row(2, "b")))
+
+      df = sql(s"SELECT * FROM $t1 WITH (`split-size` = 5)")
```

Comment:
just want to brainstorm a bit here: what should be the SQL API for `spark.readStream.table`? `WITH STREAMING options`?

Reply:
It might be closer to the SQL standard to use TVF syntax for this purpose instead (possibly with named arguments). This example would look like:

      df = sql(s"SELECT * FROM $t1(split-size => 5)")

Reply:
ok how about we add a TVF like syntax but it's not actual TVF, so that we can do `FROM BATCH(tbl, k1 => v1, ...)`, and `FROM STREAM(tbl, k1 => v1, ...)`. It's not actual TVF because the first parameter is table identifier, but we can make it a valid TVF with `FROM BATCH(TABLE tbl, k1 => v1, ...)`, which is more verbose.

Reply:
Pure TVF tableName(named parameterlist), while elegant on one level may be rather confusing to users.
There is no more visual cue what is a TVF and what is a table , and of course we would need a tie breaker for whom to resolve for first. This adds unwanted complexity.

Adding a builtin Polymorphic TVF uses existing (and standard)  language.

I would like to know some examples. Teh PR mentions federation (?).

Reply:
👍 the built-in polymorphic TVF sounds like a clear use-case here. Thanks @srielau @cloud-fan for the suggestions. ➕1️⃣ for a few examples as well to help us work together :)